src/main/java/org/truffleruby/language/arguments/RubyArguments.java

```diff
@@ -82,7 +83,14 @@
         packed[ArgumentIndicies.DECLARATION_CONTEXT.ordinal()] = declarationContext;
         packed[ArgumentIndicies.FRAME_ON_STACK_MARKER.ordinal()] = frameOnStackMarker;
         packed[ArgumentIndicies.SELF.ordinal()] = self;
-        packed[ArgumentIndicies.BLOCK.ordinal()] = block;
+
+        /* The block in the arguments array is always either a Nil or RubyProc. The conversion from null if the caller
+         * doesn't want to provide a block is done at the caller, because it will know the type of values within its
+         * compilation unit.
+         *
+         * When you read the block back out in the callee, you'll therefore get a Nil or RubyProc. */
+        assert block == null || block instanceof Nil || block instanceof RubyProc : block;
```

Comment:
There are like a dozen callers of RubyArguments.pack(), so I think it's actually quite easy to modify them to pass either a RubyProc or Nil, and never null (and adapt the `assert`), so I would suggest to do it in this PR.

Reply:
Agreed. I’ll do that now.

Reply:
I’ve made this change in the most conservative way possible, i.e. replace any literal `null`s in the callers with `nil`s, and do null checks everywhere else. This is still not perfect but should result in even fewer null checks.

I think we can remove most (if not all) of these replacement null checks in a future PR but it requires a fair amount of chasing up the stack and changing more method signatures to refer to `Object` rather than `RubyProc` so I’d rather not do it here.